### PR TITLE
Implement OneULP and related ULP functions

### LINKED ISSUE #1148: Change how ULP threshholds are calculated
For a constraint that is described as `c` ULP in the spec, where `c` is a constant value; currently the CTS is implementing a check that the generated values is within `c` ULP steps of the truth value, i.e. the calculated result's binary representation is within `c` bits of the expected value.

Closer reading of the spec indicates that the intended accuracy is actually `c` times one ULP step at the expected value. For the vast majority of cases, these two interpretations are equivalent, except when the expected value is near the boundary in the floating point representation where the difference neighbouring values have different exponents. This means that the tests as written is are more strict than intended.

Vulkan describes their constraints and implements them as `c` * 1ULP, so I am confident that this is the intended reading, since WGSL should have tighter constraints then one of the backend platforms.

### PR DESCRIPTION
**Pull request by zoddicus**
Replaces the existing ULP accuracy checking code

Fixes #1148

<hr>

**Requirements for PR author:**

- [x] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [x] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [x] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

**Review by dj2 — COMMENTED** [review]
typescript lgtm modulo one nit. I don't have enough knowledge to say if the various test results are accurate.

File: src/unittests/maths.spec.ts [hunk 1]
```diff
@@ -217,7 +161,17 @@ g.test('test,math,nextAfterNoFlush')
     { val: hexToF32(0x03800000), dir: true, result: f32Bits(0x03800001) },
     { val: hexToF32(0x03800000), dir: false, result: f32Bits(0x037fffff) },
     { val: hexToF32(0x83800000), dir: true, result: f32Bits(0x837fffff) },
-    { val: hexToF32(0x83800000), dir: false, result: f32Bits(0x83800001) },
+   { val: hexToF32(0x83800000), dir: false, result: f32Bits(0x83800001) },
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/unittests/maths.spec.ts]: Done

File: src/webgpu/util/math.ts [hunk 2]
```diff
@@ -149,29 +118,98 @@ export function nextAfter(val: number, dir: boolean = true, flush: boolean): Sca
...
+
+  // For values at the edge of the range or beyond ulp(x) is defined as
+  // the distance between the two nearest representable numbers to the
+  // appropriate edge.
+  if (target === Number.POSITIVE_INFINITY || target >= kValue.f32.positive.max) {
+    return kValue.f32.positive.max - kValue.f32.positive.nearest_max;
+  } else if (target === Number.NEGATIVE_INFINITY || target <= kValue.f32.negative.min) {
+    return kValue.f32.negative.nearest_min - kValue.f32.negative.min;
+  } else {
+    const b = nextAfter(target, true, flush).value.valueOf() as number;
```
Comment (ben-clayton) [inline src/webgpu/util/math.ts:182]: As discussed, I'm not sure about this - it seems like the ULP is (nearly always) twice what it should be.

File: src/unittests/maths.spec.ts [hunk 3]
```diff
@@ -36,85 +37,18 @@ function hexToFloat64(h32: number, l32: number): number {
 /**
  * @returns true if arrays are equal, doing element-wise comparison as needed, and considering NaNs to be equal.
  *
- * Depends on the correctness of diffULP, which is tested in this file.
+ * Depends on the correctness of withinULP, which is tested in this file.
```
[the remark this answers is not in the document]
Comment (zoddicus) [reply src/unittests/maths.spec.ts:40]: Done

File: src/webgpu/util/math.ts (hunk 2 above)
Comment (zoddicus) [reply src/webgpu/util/math.ts]: Done

**Comment by zoddicus** [comment]
@ben-clayton PTAL

**Review by ben-clayton — APPROVED** [review]
Looks solid. A request for a few more cases.

**Pull request merged**